### ISSUE TITLE
ORM requires update due to latest Doorkeeper changes

### ISSUE THREAD

**Issue by nbulaj** [nbulaj/doorkeeper-sequel#33]
Doorkeeper has ORM changes merged to `master` branch.

See https://github.com/doorkeeper-gem/doorkeeper/commit/25d038022c2fcad45af5b73f9d003cf38ff491f6 for the details.

**Comment by nbulaj** [comment]
@gencer maybe you wanna to look at it? If no, I'll do it my own.
This is a security related changes

**Comment by gencer** [comment]
I just pull `master` branch of `doorkeeper` gem and trigger a fresh build. I got lot of errors regarding to ActiveRecord. No modification made yet.

https://travis-ci.org/github/gencer/doorkeeper-sequel/jobs/682542459#L758


Did something changed on doorkeeper other than this security change?

ORM is Sequel but it forcibly trying to load ActiveRecord.


```
      NameError:
        uninitialized constant ActiveRecord
      # ./doorkeeper/lib/doorkeeper/orm/active_record/access_grant.rb:6:in `<module:Doorkeeper>'
      # ./doorkeeper/lib/doorkeeper/orm/active_record/access_grant.rb:5:in `<top (required)>'
      # ./doorkeeper/lib/doorkeeper/orm/active_record.rb:19:in `block in initialize_models!'
      # ./doorkeeper/lib/doorkeeper/orm/active_record.rb:46:in `lazy_load'
      # ./doorkeeper/lib/doorkeeper/orm/active_record.rb:17:in `initialize_models!'
      # ./doorkeeper/lib/doorkeeper/config.rb:50:in `setup_orm_models'
      # ./doorkeeper/lib/doorkeeper/config.rb:26:in `configure'
      # ./spec/controllers/tokens_controller_spec.rb:10:in `block (2 levels) in <top (required)>'
      # ./spec/spec_helper.rb:43:in `block (3 levels) in <top (required)>'
      # ./spec/spec_helper.rb:43:in `block (2 levels) in <top (required)>'
      # ./spec/spec_helper_integration.rb:43:in `block (3 levels) in <top (required)>'
      # ./spec/spec_helper_integration.rb:43:in `block (2 levels) in <top (required)>'

```

Before I start to work on this, I want to make sure of this change.

**Comment by nbulaj** [comment]
Ah, nothing changed I hope. I'll check when I have time, thanks!

**Comment by gencer** [comment]
@nbulaj `CVE-2020-10187` seems very critical. Can we work on this as soon as possible?

**Comment by nbulaj** [comment]
I have the same issue. Need to check what changed to understand why it requires ActiveRecord

https://github.com/doorkeeper-gem/doorkeeper/compare/17017ec9b18c781aec74ac798cad2d7ed9fc72f9..b275c6aaf5bddc7af38409835de3161bc32ee19f

**Comment by nbulaj** [comment]
Hmm, this one broke https://github.com/doorkeeper-gem/doorkeeper/commit/de7cdc66f7053e0ef4b69a361d2a9ffd0963358e

But I don't understand why AR ORM `initialize_models!` method called :thinking: It must use Sequel instead O_o

**Comment by nbulaj** [comment]
OK, fixed with doorkeeper master. Now sequel model must be updated

**Comment by gencer** [comment]
The thing is `master` branch of doorkeeper relies on AR's serializer and there is a method called `read_attribute_for_serialization`. This enables inclusion of `secret` when requested and exclude when not needed.

I have to find a way to do this with Sequel.

**Comment by gencer** [comment]
I tried to look your `mongodb` gem but it seems you did not update there yet. I am missing something here.

When I look `as_json` from AR, I see that it is actually a serializable hash. Which is similar to json serialization, we can `exclude`, `include` and use `only` there.

However, Sequel has no such feature. And I don't want to import/use AR specific class in here.

See: https://api.rubyonrails.org/classes/ActiveModel/Serialization.html

Have you find anything for MongoDB yet? So that i can use it in here? Or any helper method that i don't know yet?

**Comment by nbulaj** [comment]
How about https://sequel.jeremyevans.net/rdoc-plugins/classes/Sequel/Plugins/JsonSerializer.html ?

**Comment by gencer** [comment]
`to_json` returns string. `as_json` however expected to return Hash. Thus, specs compare via `match`.

We have two options:

a. Only use `to_json` and skip `as_json` on tests
b. Find a way to serialize hash

`as_json` returns as hash, isn't it? Am I wrong? Because, in that method if I return `json`, rspec test fail with `expected Hash`. If I return Hash then compare result.

Even in mongodb gem it returns hash: https://github.com/doorkeeper-gem/doorkeeper-mongodb/blob/d9426a5d7ad3143a59082cb04f64f9b9323a920e/lib/doorkeeper-mongodb/mixins/mongoid/application_mixin.rb#L155

Also see here how we serialize and filter hash on `as_json`: https://github.com/doorkeeper-gem/doorkeeper/commit/25d038022c2fcad45af5b73f9d003cf38ff491f6#diff-db8e1aa900c1af4541224aa0127ae853L68

**Comment by gencer** [comment]
@nbulaj Should we copy/adopt `ActiveModel::Serialization` into `DoorkeeperSequel` or should we write our own serializer? Have you find anything for `mongodb` so that i can look into it?

`as_json` returns Hash. There are operations in this method that excludes or includes keys by given `options`. Sequel has no such feature. Sequel only has `to_json`. Since, `to_json` capable of exclude things, it doesn't return `Hash`.

**Comment by nbulaj** [comment]
Hi @gencer. Sorry, I don't have much time for it now.

I don't sure about best solution here. We could write our own `to_json` implementation, but it wouldn't work like users expect it to be (with all the options, considering expressions and so on). By the other hand it's very hard to reuse to reuse https://github.com/jeremyevans/sequel/blob/master/lib/sequel/plugins/json_serializer.rb just because we need either totally copy `#to_json` implementation or do some tricks like

```ruby
h = super
JSON.parse(h)
h.remove("secret")
JSON.dump(h)
```

Which is also a :hankey: 

For mongodb I think I can just reuse ActiveModel::Serialization, but I don't sure it fits to Sequel :(

**Comment by nbulaj** [comment]
Actually I think we can play with `except:` option to skip sensitive columns

Something like

```ruby
def to_json(*options)
  if ...
   super(options.merge(except: ...)
  end
end
```

**Comment by gencer** [comment]
@nbulaj I also didn't like the `JSON.parse` idea, however, until I write a serializer for doorkeeper-sequel, we will live with it i guess.

Here you go: https://travis-ci.org/github/gencer/doorkeeper-sequel/builds/686186227

All success!

**Comment by nbulaj** [comment]
Released as 2.3
